### PR TITLE
[Type checker] Introduce a request to evaluate “where” clause requirements

### PR DESCRIPTION
**Pull request by DougGregor**
Introduce a new form of request that produces the ith Requirement of the
where clause of a given entity, which could be a protocol, associated type,
or anything with generic parameters. The requirement can be evaluated
at any type resolution stage; the “interface” type stage will be cached
in the existing RequirementReprs.

Fixes SR-8119 / rdar://problem/41498944.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test compiler performance

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Review by slavapestov — APPROVED** [review]
Great progress here

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -208,28 +208,7 @@ class DenseMapInfo<RawValueKey> {
...
-
-  for (auto assocType : protocol->getAssociatedTypeMembers()) {
-    if (auto whereClause = assocType->getTrailingWhereClause()) {
-      revertGenericRequirements(whereClause->getRequirements());
-      validateRequirements(whereClause->getWhereLoc(),
-                           whereClause->getRequirements(),
-                           resolution, options);
-    }
-  }
+  // FIXME: DELETE
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply lib/Sema/TypeCheckDecl.cpp]: It's in the next commit. I didn't want to rebuild the world yet ;)

File: lib/Sema/TypeCheckDecl.cpp [hunk 2]
```diff
@@ -536,14 +515,24 @@ static void checkInheritanceClause(
...
     return;
 
-  for (auto gp : *genericParams)
+  for (auto gp : *genericParams) {
     checkInheritanceClause(gp);
+  }
+
+  // Force visitation of each of the requirements here.
+  auto owner =
+    genericParams->getParams().front()->getDeclContext()->getAsDecl();
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply lib/Sema/TypeCheckDecl.cpp]: Yeah, I'll pass it down.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test compiler performance

**Pull request merged**